File: HashTable/HashTableLP.java

public class HashTableLP {
    private static int EMPTY_VALUE = -1;
    private static int DELETED_VALUE = -2;
    private static int FILLED_VALUE = 0;

    private int tableSize;
    int[] array;
    int[] Flag;

    public HashTableLP(int tSize) {
        tableSize = tSize;
        array = new int[tSize + 1];
        Flag = new int[tSize + 1];
        for (int i = 0; i <= tSize; i++) {
            Flag[i] = EMPTY_VALUE;
        }
    }

    /* Other Methods */

    int computeHash(int key) {
        return key % tableSize;
    }
    
    int resolverFun(int index) {
        return index;
    }
    
    int resolverFun2(int index) {
        return index * index;
    }
    
    boolean add(int value) {
        int hashValue = computeHash(value);
        for (int i = 0; i < tableSize; i++) {
            if (Flag[hashValue] == EMPTY_VALUE || Flag[hashValue] == DELETED_VALUE) {
                array[hashValue] = value;
                Flag[hashValue] = FILLED_VALUE;
                return true;
            }
            hashValue += resolverFun(i);
            hashValue %= tableSize;
        }
        return false;
    }
    
    boolean find(int value) {
        int hashValue = computeHash(value);
        for (int i = 0; i < tableSize; i++) {
            if (Flag[hashValue] == EMPTY_VALUE) {
                return false;
            }
    
            if (Flag[hashValue] == FILLED_VALUE && array[hashValue] == value) {
                return true;
            }
    
            hashValue += resolverFun(i);
            hashValue %= tableSize;
        }
        return false;
    }
    
    boolean remove(int value) {
        int hashValue = computeHash(value);
        for (int i = 0; i < tableSize; i++) {
            if (Flag[hashValue] == EMPTY_VALUE) {
                return false;
            }
    
            if (Flag[hashValue] == FILLED_VALUE && array[hashValue] == value) {
                Flag[hashValue] = DELETED_VALUE;
                return true;
            }
            hashValue += resolverFun(i);
            hashValue %= tableSize;
        }
        return false;
    }
    
    void print() {
        System.out.print("Hash Table contains ::");
        for (int i = 0; i < tableSize; i++) {
            if (Flag[i] == FILLED_VALUE) {
                System.out.print(array[i] + " ");
            }
        }
        System.out.println();
    }
    
    public static void main(String[] args) {
        HashTableLP ht = new HashTableLP(1000);
        ht.add(1);
        ht.add(2);
        ht.add(3);
        ht.print();
        System.out.println("Find key 2 : " + ht.find(2));
        ht.remove(2);
        System.out.println("Find key 2 : " + ht.find(2));
    }
}

/*
Hash Table contains ::1 2 3 
Find key 2 : true
Find key 2 : false
*/

File: HashTable/HashTableLP.js
var EMPTY_VALUE = 0;
var FILLED_VALUE = 1;
var DELETED_VALUE = 2;
var HashTableLP = /** @class */ (function () {
    function HashTableLP(tSize) {
        if (this.tableSize === undefined)
            this.tableSize = 0;
        if (this.Arr === undefined)
            this.Arr = null;
        if (this.Flag === undefined)
            this.Flag = null;
        this.tableSize = tSize;
        this.Arr = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(tSize + 1);
        this.Flag = (function (s) { var a = []; while (s-- > 0)
            a.push(0); return a; })(tSize + 1);
        for (var i = 0; i <= tSize; i++) {
            {
                this.Flag[i] = EMPTY_VALUE;
            }
            ;
        }
    }
    HashTableLP.prototype.computeHash = function (key) {
        return key % this.tableSize;
    };
    HashTableLP.prototype.resolverFun = function (index) {
        return index;
    };
    HashTableLP.prototype.resolverFun2 = function (index) {
        return index * index;
    };
    HashTableLP.prototype.add = function (value) {
        var hashValue = this.computeHash(value);
        for (var i = 0; i < this.tableSize; i++) {
            {
                if (this.Flag[hashValue] === EMPTY_VALUE ||
                    this.Flag[hashValue] === DELETED_VALUE) {
                    this.Arr[hashValue] = value;
                    this.Flag[hashValue] = FILLED_VALUE;
                    return true;
                }
                hashValue += this.resolverFun(i);
                hashValue %= this.tableSize;
            }
            ;
        }
        return false;
    };
    HashTableLP.prototype.find = function (value) {
        var hashValue = this.computeHash(value);
        for (var i = 0; i < this.tableSize; i++) {
            {
                if (this.Flag[hashValue] === EMPTY_VALUE) {
                    return false;
                }
                if (this.Flag[hashValue] === FILLED_VALUE && this.Arr[hashValue] === value) {
                    return true;
                }
                hashValue += this.resolverFun(i);
                hashValue %= this.tableSize;
            }
            ;
        }
        return false;
    };
    HashTableLP.prototype.remove = function (value) {
        var hashValue = this.computeHash(value);
        for (var i = 0; i < this.tableSize; i++) {
            {
                if (this.Flag[hashValue] === EMPTY_VALUE) {
                    return false;
                }
                if (this.Flag[hashValue] === FILLED_VALUE && this.Arr[hashValue] === value) {
                    this.Flag[hashValue] = DELETED_VALUE;
                    return true;
                }
                hashValue += this.resolverFun(i);
                hashValue %= this.tableSize;
            }
            ;
        }
        return false;
    };
    HashTableLP.prototype.print = function () {
        for (var i = 0; i < this.tableSize; i++) {
            {
                if (this.Flag[i] === FILLED_VALUE) {
                    console.info("Node at index [" + i + " ] :: " + this.Arr[i]);
                }
            }
            ;
        }
    };
    HashTableLP.main = function (args) {
        var ht = new HashTableLP(1000);
        ht.add(1);
        ht.add(2);
        ht.add(3);
        ht.print();
        console.info(ht.remove(1));
        console.info(ht.remove(4));
        ht.print();
    };
    return HashTableLP;
}());
HashTableLP["__class"] = "HashTableLP";
HashTableLP.main(null);
